### ISSUE TITLE
Thoughts

### ISSUE THREAD

**Issue by creecros** [vistree/kanboard-backlog#1]
_Then create a board and name the first (!) swimlane "backlog" and the first (!) column also "backlog"._
* I guess the goal here would be to make it so there is no need for this?
  * Just use the first column, regardless of name. Or create Backlog column if doesn't exist.
  * Create a swimlane through the plugin, eliminate the prep work for the user.

_If you change an existing board, make sure that there are only tasks in the first column in the first swimlane and that there are no tasks in the first column in all other swimlanes._
* Maybe we can figure out a way to just show everything in the column and ignore whatever swimlane it may have been in.
* If we create a swimlane to use as the first swimlane, there will be no need to worry about whether tasks were in the swimlane or not

_Most of it works. The only problem is that this way collapsing the columns does not work completely for the main area (columns that are NOT the backlog) and not at all for my new column backlog._ 
* it all works for me, I say bravo. unless I am missing something. Every column and siwmlane is collapsible.

I'd also add some type of an on/off switch on the board, by doing this, you will be able to go through your own controller where you do all the above, preventing a complete override of existing controller.

**Comment by vistree** [comment]
> Then create a board and name the first (!) swimlane "backlog" and the first (!) column also "backlog".
- you are right - we need to automate! 
- [ ] maybe best to always create new column if column[0][name] != _backlog ?!
- [ ] maybe best to always create new swimlane if swimlane[0][name] != _backlog ?!

> If you change an existing board, make sure that there are only tasks in the first column in the first swimlane and that there are no tasks in the first column in all other swimlanes.
- If we implement above workflow in always creating new smimlane/backlog, we only need to worry if user disables plugin, creates new tasks in _backlog-swimlane[columns] > 0 or other-swimlanes[column] = 0 
- [ ] on plugin-enable check, if column[0][name] = _backlog and swimlane[0][name] = _backlog
- [ ] -> if this is true, move all tasks in all columns of swimlane[0] to first column (_backlog[column][0]]) AND move all tasks in all swimlanes[column][0] also to _backlog[column][0]

> Most of it works. The only problem is that this way collapsing the columns does not work completely for the main area (columns that are NOT the backlog) and not at all for my new column backlog.
- Yes, collapsing the columns itself work, but the width of the columns after collapsing should be minimised. Currently the width stays the same in the backlog-column regardless the column is collapsed or not and it seems that the width is not optimal lowered for the other columns (compared with the original board)
- [ ] make <td> of _backlog-column to automatically resize in width after collapsing / uncollapsing AND append the free space to the other columns
- [ ] check if column size of other columns is optimised after collapsing them

> I'd also add some type of an on/off switch on the board, by doing this, you will be able to go through your own controller where you do all the above, preventing a complete override of existing controller.
- I agree! Then users can decide to enable backlog column for single project and we have a flag we can use in the first topic of this list to check against

I think, the only real unclear point is whether we need an new first column or not. 

Besides this, we can discuss if my approach in creating an extra table (which requires a unique swimlane AND a unique column) is the best, or if trying to use the first column in combination with rowspan would be the better way. As I described in the old issue, I tried this, but found it more difficult to make it work then using the extra table.

Can we convert the topics to "working" packages so we can discuss them in detail? Is it best to create an issue for each of them or is the project area the better place?

**Comment by vistree** [comment]
Great, I recognised your new branch. Anything you want me already to test?

**Comment by creecros** [comment]
Test away. I think I left off at the last function in the controller, moving the tasks out of the column before removing it. Everything else was working, although I didn't get a chance to implement into your template either.

**Comment by creecros** [comment]
Ok, everything added works, so far. Test out and let me know thoughts before I go any further.

**Comment by vistree** [comment]
Will do - thanx!

**Comment by vistree** [comment]
Hi, great work!!! Like the way you implemented the backlog switcher! I found some smaller problems with PHP5 and PHP7.1. Do you want me to change the code or better to discuss them first here in the issue?

**Comment by creecros** [comment]
No worries, just change it

**Comment by creecros** [comment]
These should have been `== 2` not `=== 2`

https://github.com/vistree/kanboard-backlog/blob/5886746b623d07dbfc879d812208f8ffa1e64283/Controller/BacklogBoardController.php#L93-L95

**Comment by creecros** [comment]
Also, added a UI prevention:
* when backlog board is set, 'Backlog_Board' column and 'Backlog_Swimlane' swimlane will be hidden from column/swimlane config.
* users can still drag and sort other columns/swimlanes, but the created swimlane/column will be hidden and will remain in first positions always.
* API calls will  be able to move them though.

**Comment by vistree** [comment]
Really good! Many thanks for the enormous progress! I will have time on Sunday to do my tasks!

**Comment by creecros** [comment]
Take your time, it's your plugin! I'm in no rush over here.

**Comment by creecros** [comment]
I'm gonna go ahead and merge where it's at, rebranch from here on out and merge when complete.

**Comment by creecros** [comment]
Anything else on this one? Think we covered all the bases. Do you know how to add to https://github.com/kanboard/website?

**Comment by vistree** [comment]
I think there is one thing left - if the backlog related column/swimlane name exists before switching to backlog-view, tasks being in the backlog-swimlane and NOT being in the backlog column or being in the backlog-columns of other swimlanes then the backlog-swimlane are hidden after switching to the backlog-view. + I think we should discuss if we really want to remove the backlog-column / -swimlane if we disable the backlog-view, or if we better leave them and leave the tasks in first backlog-swimlane/column as they are and let user clean it up if they really want to remove the extra swimlane / column.

**Comment by creecros** [comment]
The swimlane has about a 0.00001% chance of anything ever being in it, so that can be deleted on disable, leaving the column, is entirely up to you, but the code will have to be rewritten ibelieve, because now we just increased the chance of "Backlog_Board" column already exists on Enable by a factor of 1000x, i'm just making these numbers up, but the chance of it existing before enabled is now high.

**Comment by creecros** [comment]
also, if you don't delete it, there is the posibility that enable/disable could be repeated to the point of a whole bunch of columns being created and not removed.

**Comment by creecros** [comment]
didn't meant to close. pushed wrong button!

**Comment by vistree** [comment]
You are right - let's go with the concept and see if we get any feedback from other users ;-)

**Comment by creecros** [comment]
K, i'll add to my current pull request kanboard/website#89

**Comment by creecros** [comment]
sent you a collaborator invite on another plugin i was working on, if your interested
